### PR TITLE
`std.os.linux.start_pie`: Add mips and powerpc support

### PR DESCRIPTION
**Pull request by alexrp**
Requires #20822 to actually work.

### CODE REVIEW

File: lib/std/os/linux/start_pie.zig [hunk 1]
```diff
@@ -111,6 +113,27 @@ fn getDynamicSymbol() [*]elf.Dyn {
...
+            \\ .gpword _DYNAMIC
+            \\ 1:
+            \\ lw %[ret], 0($ra)
+            \\ addu %[ret], %[ret], $gp
+            : [ret] "=r" (-> [*]elf.Dyn),
+        ),
+        .mips64, .mips64el => asm volatile (
+            \\ .weak _DYNAMIC
+            \\ .hidden _DYNAMIC
+            \\ .balign 8
```
Comment (alexrp) [inline lib/std/os/linux/start_pie.zig:131]: https://github.com/ziglang/zig/pull/20822#discussion_r1693893040

**Comment by alexrp** [comment]
cc @kubkon you might want to check the relocation stuff?

File: lib/std/os/linux/start_pie.zig [hunk 2]
```diff
@@ -149,34 +205,61 @@ pub fn relocate(phdrs: []elf.Phdr) void {
...
-        const rel = std.mem.bytesAsSlice(elf.Rel, @as([*]u8, @ptrFromInt(rel_addr))[0..rel_size]);
-        for (rel) |r| {
+    // Deal with the GOT relocations that MIPS uses first.
+    if (builtin.cpu.arch.isMIPS()) {
+        var count: elf.Addr = 0;
+
+        // This is an architecture-specific tag, so not part of `sorted_dynv`.
+        var j: usize = 0;
+        while (dynv[j].d_tag != elf.DT_NULL) : (j += 1) {
+            if (dynv[j].d_tag == elf.DT_MIPS_LOCAL_GOTNO) count = dynv[j].d_val;
```
[the remark this answers is not in the document]
Comment (alexrp) [reply lib/std/os/linux/start_pie.zig:232]: This part mirrors what musl does. I found it odd too. I don't suppose you have any idea if this tag is allowed to appear multiple times?

[the remark this answers is not in the document]
Comment (kubkon) [reply lib/std/os/linux/start_pie.zig:232]: Given that `DT_MIPS_LOCAL_GOTNO` records a count of local GOT entries that need rebasing, you can either `break` early, or do `@max()`. Having said that if the linker emits more than one `DT_MIPS_LOCAL_GOTNO` entry with differing values it's cleary an ill-formed binary in which case we could even `@trap`.

[the remark this answers is not in the document]
Comment (alexrp) [reply lib/std/os/linux/start_pie.zig]: Ok, I changed it to just take the first one we find. I figure if that's somehow not appropriate in some situations, the breakage will be fairly loud (early/reliable crashes).

**Review by kubkon — APPROVED** [review]

**Comment by andrewrk** [comment]
Nice!

**Pull request merged**